Fix the overflow in `CircularBuffer::set_capacity`

Krebrov001 opened the pull request:
- Relates to #2507 
- Fixes #2507 

 - Previously the loop that used to copy the newest elements had a wrong offset.
 - It didn't correctly take into account looping over the `m_head` in the modulo.
 - I refactored the code to have a simplified loop using a proven class method.

Krebrov001 committed:
Replace `[]` with `.at()` in `m_buffer` for ease of error detection

 - Previously one-off errors would silently overwrite memory, and the tests would pass.
 - Now the tests fail when an incorrect index gets accessed.

Signed-off-by: Konstantin Rebrov <konstantin.rebrov@intel.com>

Krebrov001 committed:
Fix the overflow in `CircularBuffer::set_capacity`

 - Previously the loop that used to copy the newest elements had a wrong offset.
 - It didn't correctly take into account looping over the `m_head` in the modulo.
 - I refactored the code to have a simplified loop using a proven class method.

Signed-off-by: Konstantin Rebrov <konstantin.rebrov@intel.com>

dannosliwcd reviewed (changes requested):
The implementation change looks good. Please update the doxygen strings and add a unit test case. 

* Document that `set_capacity()` retains the newest elements when it is set to a value smaller than the current size.
* Document that `value()` accepts negative indices to represent nth-most-recent insertions. Our documentation currently calls that an error, but this change explicitly depends on that behavior not being treated as an error.
* Add a unit test that verifies that `value()` correctly wraps negative inputs.

Krebrov001 committed:
Update `CircularBuffer` methods

- Make the doxygen comments more detailed for `set_capacity()`
- Make `value()` accept negative indices to represent nth-most-recent insertions.
- Update the doxygen comments to reflect this change in `value()` method.
- Add a unit test that verifies that `value()` correctly wraps negative inputs.

Signed-off-by: Konstantin Rebrov <konstantin.rebrov@intel.com>

Krebrov001 committed:
Update the man pages to reflect changes to `CircularBuffer` methods

- Make the contents of the man pages match what's in the doxygen comments.

Signed-off-by: Konstantin Rebrov <konstantin.rebrov@intel.com>

dannosliwcd commented on service/src/geopm/CircularBuffer.hpp:
```diff
@@ -69,17 +72,23 @@ namespace geopm
             ///
             /// @return Value from the specified buffer index.
-            const type& value(const unsigned int index) const;
+            const type& value(const int index) const;
```
@cmcantalupo I think this type change is allowed for 2.1 since this is a header-only interface (users always recompile with it so we don't need to worry about caller/callee type mismatches). Is that true?

Krebrov001 replied:
It's ok to make the change in this case.

dannosliwcd commented on service/src/geopm/CircularBuffer.hpp:
```diff
@@ -69,17 +72,23 @@ namespace geopm
-            /// of valid entries in the buffer. An attempt to
-            /// retrieve a value for an out of bound index
+            /// at a particular index.
+            /// There are two kinds of valid indices: positive and negative ones.
```
```suggestion
```
Positive,  negative, _and zero_ are allowed. Since `const int index` already limits to that domain, you can omit this line.

dannosliwcd commented on service/src/geopm/CircularBuffer.hpp:
```diff
@@ -69,17 +72,23 @@ namespace geopm
+            /// Where size is the number of valid entries in the buffer.
+            /// An out of bounds index is > [size-1] OR < [-size].
+            /// An attempt to retrieve a value for an out of bound index
             /// will throw a geopm::Exception with an
```
This level of detail is great! I think a user will know exactly what should/shouldn't go in to this function, and what will happen for each class of good and bad inputs.

dannosliwcd commented on service/src/geopm/CircularBuffer.hpp:
```diff
@@ -176,22 +185,28 @@ namespace geopm
+    const type& CircularBuffer<type>::value(const int index) const
     {
-        if (index >= m_count) {
+        if (index >= 0 && index >= m_count ||
+            index <  0 && index < -m_count) {
```
```suggestion
        if (index >= m_count || index < -m_count) {
```

`unsigned long m_count` cannot be negative (by data type and by nature of being a non-borrowable count), so `index >= m_count` implies `index >= 0` and `index < -m_count` implies `index < 0`.

Krebrov001 replied:
That change causes a lot of tests to fail, so it cannot be implemented.

dannosliwcd commented on service/src/geopm/CircularBuffer.hpp:
```diff
@@ -176,22 +185,28 @@ namespace geopm
             throw Exception("CircularBuffer::value(): index is out of bounds", GEOPM_ERROR_INVALID, __FILE__, __LINE__);
         }
-        return m_buffer[(m_head + index) % m_max_size];
+        if (index < 0) {
+            const int new_index = m_count + index;
+            return m_buffer.at((m_head + new_index) % m_max_size);
+        } else {
+            return m_buffer.at((m_head + index) % m_max_size);
+        }
```
Would it be sufficient to only do the single following case?

`return m_buffer.at((m_head + index) % m_count)`

I think that would work as long as we guarantee `m_count` is always `<= m_max_size`. Thoughts?

Krebrov001 replied:
@dannosliwcd That doesn't work. I already tried it.

Krebrov001 committed:
Addressing review feedbacks

Signed-off-by: Konstantin Rebrov <konstantin.rebrov@intel.com>
